## Naive Benchmarks for VirulencePredictor

Using different models to train on the Host-Pathogen Virulence Prediction dataset. This would create benchmarks to use in the GCMC model.

None of these models take in account the underlying graph structure of the model, and thus don't have very high performance.

All the models are traditional Machine Learning Regression models.

**Imports**

In [1]:
import pickle
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings('ignore')

from sklearn.preprocessing import MinMaxScaler

**Creating the dataframe**

In [2]:
# Collect data from the created features for virus and mice.

data_repo = 'data/'

with open(data_repo + 'graph_info.pkl') as f:
    graph_info = pickle.load(f)
num_users, num_items, u_nodes, v_nodes, ratings = graph_info

with open(data_repo + 'virus_enc_ae.pkl') as f:
    u_features = pickle.load(f)

with open(data_repo + 'mouse_enc_pca.pkl') as f:
    v_features = pickle.load(f)

In [3]:
# The sample is created by concatenating the corresponding
# virus and mouse nodes in the graph

X, y = [], []
for i in range(len(u_nodes)):
    u = u_nodes[i]
    v = v_nodes[i]

    feat_u = u_features[u]
    feat_v = v_features[v]
    
    X.append(np.concatenate((feat_u, feat_v), axis=0))
    y.append(ratings[i])
    
X = np.vstack(X)
y = np.array(y, dtype='float32')

In [4]:
X.shape, y.shape

((306, 200), (306,))

In [5]:
# Create the dataframe
headers = ['a' + str(i + 1) for i in range(X.shape[1])]
df = pd.DataFrame(X, columns=headers)

In [6]:
df.head()

a1   a2         a3         a4   a5         a6   a7   a8   a9  a10  \
0  0.0  0.0   0.000000  25.147011  0.0  22.536587  0.0  0.0  0.0  0.0   
1  0.0  0.0   0.000000  24.914425  0.0  22.352468  0.0  0.0  0.0  0.0   
2  0.0  0.0  27.604977  37.230022  0.0  19.131760  0.0  0.0  0.0  0.0   
3  0.0  0.0   0.000000  25.438612  0.0  22.710747  0.0  0.0  0.0  0.0   
4  0.0  0.0  27.606718  37.237728  0.0  19.136921  0.0  0.0  0.0  0.0   

    ...         a191      a192     a193      a194      a195      a196  \
0   ...    -3.054926 -1.107168 -1.15011 -3.535913  0.053217 -1.155408   
1   ...    -3.054926 -1.107168 -1.15011 -3.535913  0.053217 -1.155408   
2   ...    -3.054926 -1.107168 -1.15011 -3.535913  0.053217 -1.155408   
3   ...    -3.054926 -1.107168 -1.15011 -3.535913  0.053217 -1.155408   
4   ...    -3.054926 -1.107168 -1.15011 -3.535913  0.053217 -1.155408   

       a197      a198      a199     a200  
0 -2.041763 -3.035815 -2.366308 -2.81482  
1 -2.041763 -3.035815 -2.366308 -2.81482  
2 -2.041763 -3.035815 -2.366308 -2.81482  
3 -2.041763 -3.035815 -2.366308 -2.81482  
4 -2.041763 -3.035815 -2.366308 -2.81482  

[5 rows x 200 columns]

**Preprocessing**

In [7]:
# Check that no NaNs exist
assert df.isnull().sum().all() == 0

In [8]:
# Normalize the dataset
scaler = MinMaxScaler()
df = scaler.fit_transform(df)

In [9]:
df = pd.DataFrame(df, columns=headers)

In [10]:
df.head()

a1   a2        a3        a4   a5        a6   a7   a8   a9  a10    ...     \
0  0.0  0.0  0.000000  0.675290  0.0  0.987265  0.0  0.0  0.0  0.0    ...      
1  0.0  0.0  0.000000  0.669044  0.0  0.979200  0.0  0.0  0.0  0.0    ...      
2  0.0  0.0  0.133663  0.999763  0.0  0.838109  0.0  0.0  0.0  0.0    ...      
3  0.0  0.0  0.000000  0.683120  0.0  0.994895  0.0  0.0  0.0  0.0    ...      
4  0.0  0.0  0.133671  0.999970  0.0  0.838335  0.0  0.0  0.0  0.0    ...      

   a191      a192      a193  a194      a195      a196      a197  a198  \
0   0.0  0.407415  0.312794   0.0  0.488463  0.301066  0.004493   0.0   
1   0.0  0.407415  0.312794   0.0  0.488463  0.301066  0.004493   0.0   
2   0.0  0.407415  0.312794   0.0  0.488463  0.301066  0.004493   0.0   
3   0.0  0.407415  0.312794   0.0  0.488463  0.301066  0.004493   0.0   
4   0.0  0.407415  0.312794   0.0  0.488463  0.301066  0.004493   0.0   

       a199      a200  
0  0.032584  0.016502  
1  0.032584  0.016502  
2  0.032584  0.016502  
3  0.032584  0.016502  
4  0.032584  0.016502  

[5 rows x 200 columns]

In [11]:
y

array([7., 7., 7., 5., 7., 7., 7., 7., 7., 7., 7., 7., 7., 6., 6., 6., 5.,
       5., 6., 6., 3., 5., 7., 7., 7., 7., 7., 7., 4., 4., 7., 7., 4., 4.,
       4., 6., 6., 5., 5., 5., 5., 6., 6., 6., 1., 7., 7., 7., 7., 7., 7.,
       7., 7., 6., 5., 5., 4., 3., 3., 6., 3., 2., 2., 6., 6., 7., 7., 7.,
       7., 2., 2., 6., 6., 1., 1., 1., 1., 1., 4., 4., 4., 4., 3., 7., 3.,
       6., 3., 3., 3., 4., 3., 3., 3., 3., 7., 5., 7., 5., 7., 3., 6., 4.,
       2., 2., 7., 7., 7., 1., 1., 6., 4., 6., 5., 4., 4., 5., 6., 4., 4.,
       4., 3., 4., 4., 5., 8., 5., 5., 5., 4., 4., 4., 5., 5., 5., 4., 2.,
       6., 6., 3., 3., 6., 6., 5., 4., 4., 4., 4., 4., 4., 4., 5., 7., 7.,
       6., 6., 6., 6., 6., 6., 3., 5., 5., 5., 7., 6., 6., 3., 3., 3., 4.,
       3., 4., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 4., 4., 4., 4., 4., 4., 4.,
       4., 4., 4., 4., 4., 4., 4., 2., 2., 2., 5., 4., 4., 4., 3., 3., 3.,
       3., 3., 3., 3., 3.

**Training Models**

In [12]:
# Value of k for k-fold cross validation
k = 5

In [13]:
def rmse(errors):
    errors = -1 * errors
    return sum([np.sqrt(x) for x in errors])/k

**First Runs**

In [16]:
# Linear Regression
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score

lr = LinearRegression()
rmse(cross_val_score(lr, df, y, cv=k, scoring='mean_squared_error'))

21291404587.72441

In [17]:
# KernelRidgeRegression
from sklearn.kernel_ridge import KernelRidge

kr = KernelRidge(alpha=0.01, kernel='rbf')
rmse(cross_val_score(kr, df, y, cv=k, scoring='mean_squared_error'))

2.013118863841572

In [19]:
# Support Vector Regression
from sklearn.svm import SVR

svr = SVR(gamma=0.1, C=1.0, epsilon=0.2)
rmse(cross_val_score(svr, df, y, cv=k, scoring='mean_squared_error'))

2.1270876976112003

In [21]:
# Boosting (AdaBoost)
from sklearn.ensemble import AdaBoostRegressor

ab = AdaBoostRegressor(n_estimators=100)
rmse(cross_val_score(ab, df, y, cv=k, scoring='mean_squared_error'))

1.8854563121107113

In [22]:
# Boosting (Gradient Tree Boosting)
from sklearn.ensemble import GradientBoostingRegressor

gb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1,
                                max_depth=1, random_state=0, loss='ls')
rmse(cross_val_score(gb, df, y, cv=k, scoring='mean_squared_error'))

1.9554725398199864

In [25]:
# Bagging (ER Trees)
from sklearn.ensemble import ExtraTreesRegressor

ert = ExtraTreesRegressor(n_estimators=100, max_depth=None, 
                           min_samples_split=2, random_state=0)
rmse(cross_val_score(gb, df, y, cv=k, scoring='mean_squared_error'))

1.9554725398199864

In [24]:
# Bagging (Random Forest)
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(n_estimators=100)
rmse(cross_val_score(rf, df, y, cv=k, scoring='mean_squared_error'))

1.9671639451455831

**Grid Search**

In [14]:
from sklearn.model_selection import train_test_split

X = df.values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [52]:
# For SVR
from sklearn.model_selection import GridSearchCV

tuned_parameters = [{'kernel': ['rbf'], 'gamma': [1e-3, 1e-4],
                     'C': [1, 10, 100, 1000], 'epsilon': [1.0, 0.1, 1e-2, 1e-3, 1e-4]}]

clf = GridSearchCV(SVR(), tuned_parameters, cv=5,
                       scoring='neg_mean_squared_error')
clf.fit(X_train, y_train)

print clf.best_params_
print np.sqrt(-1 * min(clf.cv_results_['mean_test_score']))

{'epsilon': 1.0, 'C': 1, 'gamma': 0.001, 'kernel': 'rbf'}
1.9575814889694443


In [58]:
# For KernelRidge
tuned_parameters = [{'kernel': ['rbf'], 'gamma': [1e-2, 1e-3, 1e-4], 'alpha': [1e-2, 1e-3, 1e-4]}]

clf = GridSearchCV(KernelRidge(), tuned_parameters, cv=5,
                       scoring='neg_mean_squared_error')
clf.fit(X_train, y_train)

print clf.best_params_
print np.sqrt(-1 * min(clf.cv_results_['mean_test_score']))

{'alpha': 0.0001, 'gamma': 0.01, 'kernel': 'rbf'}
1.8394874645586412


In [60]:
tuned_parameters = [{'n_estimators': [10, 50, 100, 200, 500], 
                     'learning_rate': [1.0, 0.1, 0.01, 0.001, 0.0001]}]

clf = GridSearchCV(AdaBoostRegressor(), tuned_parameters, cv=5,
                       scoring='neg_mean_squared_error')
clf.fit(X_train, y_train)

print clf.best_params_
print np.sqrt(-1 * min(clf.cv_results_['mean_test_score']))

{'n_estimators': 100, 'learning_rate': 0.1}
1.820472360599537


In [62]:
tuned_parameters = [{'n_estimators': [10, 50, 100, 200, 500], 
                     'learning_rate': [1.0, 0.1, 0.01, 0.001, 0.0001]}]

clf = GridSearchCV(GradientBoostingRegressor(), tuned_parameters, cv=5,
                       scoring='neg_mean_squared_error')
clf.fit(X_train, y_train)

print clf.best_params_
print np.sqrt(-1 * min(clf.cv_results_['mean_test_score']))

{'n_estimators': 50, 'learning_rate': 0.1}
2.2355813049597066


In [65]:
est = [10, 50, 100, 200, 500]
dep = [1, 5, 10, 15, 20]

for e in est:
    for d in dep:
        gb = GradientBoostingRegressor(n_estimators=e, learning_rate=0.1,
                                        max_depth=d, random_state=0, loss='ls')
        print "e =", e, "d =", d, "rmse =", rmse(cross_val_score(gb, df, y, cv=k, scoring='mean_squared_error'))

e = 10 d = 1 rmse = 1.8848376584969286
e = 10 d = 5 rmse = 1.9146488046927772
e = 10 d = 10 rmse = 2.0313739451456927
e = 10 d = 15 rmse = 2.0423300364539605
e = 10 d = 20 rmse = 2.0936688982279144
e = 50 d = 1 rmse = 1.9314043703214583
e = 50 d = 5 rmse = 1.9782536527660288
e = 50 d = 10 rmse = 2.2681292760336604
e = 50 d = 15 rmse = 2.349645039290053
e = 50 d = 20 rmse = 2.4541071585753103
e = 100 d = 1 rmse = 1.9554725398199864
e = 100 d = 5 rmse = 2.036981192513548
e = 100 d = 10 rmse = 2.2775284174556902
e = 100 d = 15 rmse = 2.3552073382112577
e = 100 d = 20 rmse = 2.460136649694367
e = 200 d = 1 rmse = 1.9847053958602956
e = 200 d = 5 rmse = 2.0551582051065664
e = 200 d = 10 rmse = 2.278931573728212
e = 200 d = 15 rmse = 2.3552395231633203
e = 200 d = 20 rmse = 2.460172104826303
e = 500 d = 1 rmse = 2.01103531149264
e = 500 d = 5 rmse = 2.0634850118242767
e = 500 d = 10 rmse = 2.2789563810317306
e = 500 d = 15 rmse = 2.355240683089174
e = 500 d = 20 rmse = 2.460172798514755


In [70]:
e, d = 4, 1
gb = GradientBoostingRegressor(n_estimators=e, learning_rate=0.1,
                                max_depth=d, random_state=0, loss='ls')
print "e =", e, "d =", d, "rmse =", rmse(cross_val_score(gb, df, y, cv=k, scoring='mean_squared_error'))

e = 4 d = 1 rmse = 1.8699030913958379


In [78]:
est = [4, 5, 7, 10, 15]
dep = [1, 3, 5, 7, 9]
splits = [2, 3, 4, 5, 10]

rmses = []
for e in est:
    for d in dep:
        for s in splits:
            gb = ExtraTreesRegressor(n_estimators=e, max_depth=d, min_samples_split=s)
            r = rmse(cross_val_score(gb, df, y, cv=k, scoring='mean_squared_error'))
            print "e =", e, "d =", d, "s =", s, "rmse =", r
            rmses.append(r)
            
print min(rmses)

e = 4 d = 1 s = 2 rmse = 1.9480016390964814
e = 4 d = 1 s = 3 rmse = 1.9274873869000095
e = 4 d = 1 s = 4 rmse = 1.9486832617399614
e = 4 d = 1 s = 5 rmse = 1.9579767382141668
e = 4 d = 1 s = 10 rmse = 1.952143173943842
e = 4 d = 3 s = 2 rmse = 1.9592925553603997
e = 4 d = 3 s = 3 rmse = 1.9685703087179043
e = 4 d = 3 s = 4 rmse = 2.006243199236747
e = 4 d = 3 s = 5 rmse = 2.000917152730568
e = 4 d = 3 s = 10 rmse = 1.9992917275189304
e = 4 d = 5 s = 2 rmse = 2.0108196250062043
e = 4 d = 5 s = 3 rmse = 1.9935280766417658
e = 4 d = 5 s = 4 rmse = 2.011285311550851
e = 4 d = 5 s = 5 rmse = 1.9545189377314707
e = 4 d = 5 s = 10 rmse = 1.9718289683421528
e = 4 d = 7 s = 2 rmse = 2.1067249408007394
e = 4 d = 7 s = 3 rmse = 2.0708459826145624
e = 4 d = 7 s = 4 rmse = 2.166123750513008
e = 4 d = 7 s = 5 rmse = 2.022314580113778
e = 4 d = 7 s = 10 rmse = 2.0177875783112973
e = 4 d = 9 s = 2 rmse = 2.1716065736237544
e = 4 d = 9 s = 3 rmse = 2.2591382396435584
e = 4 d = 9 s = 4 rmse = 2.1251165

In [79]:
est = [4, 5, 7, 10, 15]
dep = [1, 3, 5, 7, 9]
splits = [2, 3, 4, 5, 10]

rmses = []
for e in est:
    for d in dep:
        for s in splits:
            rf = RandomForestRegressor(n_estimators=e, max_depth=d, min_samples_split=s)
            r = rmse(cross_val_score(rf, df, y, cv=k, scoring='mean_squared_error'))
            print "e =", e, "d =", d, "s =", s, "rmse =", r
            rmses.append(r)
            
print min(rmses)

e = 4 d = 1 s = 2 rmse = 1.9027976702805975
e = 4 d = 1 s = 3 rmse = 1.9473277028087197
e = 4 d = 1 s = 4 rmse = 1.9049136812418492
e = 4 d = 1 s = 5 rmse = 1.9410061773072538
e = 4 d = 1 s = 10 rmse = 1.9048426839715085
e = 4 d = 3 s = 2 rmse = 1.9452139133018278
e = 4 d = 3 s = 3 rmse = 2.008215288865953
e = 4 d = 3 s = 4 rmse = 1.9629702197681875
e = 4 d = 3 s = 5 rmse = 1.9569575935574264
e = 4 d = 3 s = 10 rmse = 1.9915896012410859
e = 4 d = 5 s = 2 rmse = 1.9630820232092756
e = 4 d = 5 s = 3 rmse = 2.019517181420917
e = 4 d = 5 s = 4 rmse = 1.9791532627653374
e = 4 d = 5 s = 5 rmse = 2.0243659117010986
e = 4 d = 5 s = 10 rmse = 2.0183417380750823
e = 4 d = 7 s = 2 rmse = 2.051108744523121
e = 4 d = 7 s = 3 rmse = 1.9892701526954917
e = 4 d = 7 s = 4 rmse = 2.0369864996186218
e = 4 d = 7 s = 5 rmse = 2.110893236609951
e = 4 d = 7 s = 10 rmse = 2.0323485434940443
e = 4 d = 9 s = 2 rmse = 2.0574268572762957
e = 4 d = 9 s = 3 rmse = 2.104765130829885
e = 4 d = 9 s = 4 rmse = 2.047607

In [15]:
import keras.backend as K

from keras.layers import Dense
from keras.models import Sequential

def root_mean_squared_error(y_true, y_pred):
        return K.sqrt(K.mean(K.square(y_pred - y_true), axis=-1)) 

model = Sequential()
model.add(Dense(150, input_dim=200, kernel_initializer='normal', activation='relu'))
model.add(Dense(125, kernel_initializer='normal', activation='relu'))
model.add(Dense(100, kernel_initializer='normal', activation='relu'))
model.add(Dense(75, kernel_initializer='normal', activation='relu'))
model.add(Dense(50, kernel_initializer='normal', activation='relu'))
model.add(Dense(25, kernel_initializer='normal', activation='relu'))
model.add(Dense(10, kernel_initializer='normal', activation='relu'))
model.add(Dense(5, kernel_initializer='normal', activation='relu'))
model.add(Dense(1, kernel_initializer='normal'))

# Compile model
model.compile(loss=root_mean_squared_error, optimizer='adam')
history = model.fit(X, y, batch_size=32, epochs=500, validation_split=0.2, shuffle=True)

Using TensorFlow backend.


Train on 244 samples, validate on 62 samples
Epoch 1/500
244/244 [==============================] - 1s 6ms/step - loss: 4.5252 - val_loss: 4.9916
Epoch 2/500
244/244 [==============================] - 0s 149us/step - loss: 4.5167 - val_loss: 4.9829
Epoch 3/500
244/244 [==============================] - 0s 195us/step - loss: 4.5077 - val_loss: 4.9736
Epoch 4/500
244/244 [==============================] - 0s 170us/step - loss: 4.4977 - val_loss: 4.9606
Epoch 5/500
244/244 [==============================] - 0s 176us/step - loss: 4.4700 - val_loss: 4.8818
Epoch 6/500
244/244 [==============================] - 0s 176us/step - loss: 4.1726 - val_loss: 3.9183
Epoch 7/500
244/244 [==============================] - 0s 194us/step - loss: 2.2182 - val_loss: 2.0739
Epoch 8/500
244/244 [==============================] - 0s 195us/step - loss: 1.6126 - val_loss: 2.3561
Epoch 9/500
244/244 [==============================] - 0s 198us/step - loss: 1.5265 - val_loss: 2.1219
Epoch 10/500
244/244 [========

244/244 [==============================] - 0s 137us/step - loss: 1.2792 - val_loss: 2.3794
Epoch 80/500
244/244 [==============================] - 0s 137us/step - loss: 1.2663 - val_loss: 2.3566
Epoch 81/500
244/244 [==============================] - 0s 124us/step - loss: 1.2718 - val_loss: 2.3972
Epoch 82/500
244/244 [==============================] - 0s 189us/step - loss: 1.2744 - val_loss: 2.4392
Epoch 83/500
244/244 [==============================] - 0s 177us/step - loss: 1.2929 - val_loss: 2.3003
Epoch 84/500
244/244 [==============================] - 0s 156us/step - loss: 1.2892 - val_loss: 2.4924
Epoch 85/500
244/244 [==============================] - 0s 184us/step - loss: 1.3098 - val_loss: 2.2292
Epoch 86/500
244/244 [==============================] - 0s 165us/step - loss: 1.3993 - val_loss: 2.4799
Epoch 87/500
244/244 [==============================] - 0s 161us/step - loss: 1.3165 - val_loss: 2.3148
Epoch 88/500
244/244 [==============================] - 0s 230us/step - loss:

244/244 [==============================] - 0s 197us/step - loss: 1.2614 - val_loss: 2.3701
Epoch 158/500
244/244 [==============================] - 0s 195us/step - loss: 1.2494 - val_loss: 2.3896
Epoch 159/500
244/244 [==============================] - 0s 236us/step - loss: 1.2482 - val_loss: 2.4494
Epoch 160/500
244/244 [==============================] - 0s 297us/step - loss: 1.2736 - val_loss: 2.3536
Epoch 161/500
244/244 [==============================] - 0s 199us/step - loss: 1.2900 - val_loss: 2.3621
Epoch 162/500
244/244 [==============================] - 0s 257us/step - loss: 1.2881 - val_loss: 2.3885
Epoch 163/500
244/244 [==============================] - 0s 146us/step - loss: 1.2565 - val_loss: 2.3854
Epoch 164/500
244/244 [==============================] - ETA: 0s - loss: 1.471 - 0s 161us/step - loss: 1.2522 - val_loss: 2.3984
Epoch 165/500
244/244 [==============================] - 0s 176us/step - loss: 1.2533 - val_loss: 2.4254
Epoch 166/500
244/244 [======================

Epoch 235/500
244/244 [==============================] - 0s 163us/step - loss: 1.2620 - val_loss: 2.3291
Epoch 236/500
244/244 [==============================] - 0s 162us/step - loss: 1.2561 - val_loss: 2.3646
Epoch 237/500
244/244 [==============================] - 0s 150us/step - loss: 1.2329 - val_loss: 2.3957
Epoch 238/500
244/244 [==============================] - 0s 196us/step - loss: 1.2353 - val_loss: 2.3228
Epoch 239/500
244/244 [==============================] - 0s 144us/step - loss: 1.2624 - val_loss: 2.4425
Epoch 240/500
244/244 [==============================] - 0s 144us/step - loss: 1.2405 - val_loss: 2.3826
Epoch 241/500
244/244 [==============================] - 0s 169us/step - loss: 1.2423 - val_loss: 2.3599
Epoch 242/500
244/244 [==============================] - 0s 144us/step - loss: 1.2531 - val_loss: 2.3345
Epoch 243/500
244/244 [==============================] - 0s 159us/step - loss: 1.2336 - val_loss: 2.3285
Epoch 244/500
244/244 [==============================] 

244/244 [==============================] - 0s 179us/step - loss: 1.2957 - val_loss: 2.5013
Epoch 314/500
244/244 [==============================] - 0s 166us/step - loss: 1.2384 - val_loss: 2.3289
Epoch 315/500
244/244 [==============================] - 0s 182us/step - loss: 1.2058 - val_loss: 2.4140
Epoch 316/500
244/244 [==============================] - 0s 260us/step - loss: 1.2183 - val_loss: 2.3277
Epoch 317/500
244/244 [==============================] - 0s 140us/step - loss: 1.2175 - val_loss: 2.3732
Epoch 318/500
244/244 [==============================] - 0s 226us/step - loss: 1.2272 - val_loss: 2.3474
Epoch 319/500
244/244 [==============================] - 0s 161us/step - loss: 1.2154 - val_loss: 2.3052
Epoch 320/500
244/244 [==============================] - 0s 177us/step - loss: 1.2348 - val_loss: 2.4109
Epoch 321/500
244/244 [==============================] - 0s 152us/step - loss: 1.2280 - val_loss: 2.3376
Epoch 322/500
244/244 [==============================] - 0s 182us/ste

Epoch 391/500
244/244 [==============================] - 0s 155us/step - loss: 1.2145 - val_loss: 2.3303
Epoch 392/500
244/244 [==============================] - 0s 174us/step - loss: 1.2123 - val_loss: 2.3510
Epoch 393/500
244/244 [==============================] - 0s 226us/step - loss: 1.1962 - val_loss: 2.3399
Epoch 394/500
244/244 [==============================] - 0s 271us/step - loss: 1.2095 - val_loss: 2.3127
Epoch 395/500
244/244 [==============================] - 0s 215us/step - loss: 1.2145 - val_loss: 2.3606
Epoch 396/500
244/244 [==============================] - 0s 228us/step - loss: 1.1977 - val_loss: 2.3476
Epoch 397/500
244/244 [==============================] - 0s 152us/step - loss: 1.2065 - val_loss: 2.3345
Epoch 398/500
244/244 [==============================] - ETA: 0s - loss: 1.188 - 0s 182us/step - loss: 1.2160 - val_loss: 2.3778
Epoch 399/500
244/244 [==============================] - 0s 158us/step - loss: 1.2173 - val_loss: 2.4512
Epoch 400/500
244/244 [========

Epoch 469/500
244/244 [==============================] - 0s 170us/step - loss: 1.2283 - val_loss: 2.3881
Epoch 470/500
244/244 [==============================] - 0s 207us/step - loss: 1.2114 - val_loss: 2.3794
Epoch 471/500
244/244 [==============================] - 0s 137us/step - loss: 1.2270 - val_loss: 2.3070
Epoch 472/500
244/244 [==============================] - 0s 157us/step - loss: 1.2306 - val_loss: 2.5042
Epoch 473/500
244/244 [==============================] - 0s 132us/step - loss: 1.2575 - val_loss: 2.3041
Epoch 474/500
244/244 [==============================] - 0s 173us/step - loss: 1.2189 - val_loss: 2.3709
Epoch 475/500
244/244 [==============================] - 0s 179us/step - loss: 1.2039 - val_loss: 2.3499
Epoch 476/500
244/244 [==============================] - 0s 156us/step - loss: 1.2063 - val_loss: 2.3709
Epoch 477/500
244/244 [==============================] - 0s 145us/step - loss: 1.2187 - val_loss: 2.4061
Epoch 478/500
244/244 [==============================] 

In [16]:
print min(history.history['val_loss'])

2.0738949314240487


**Final Readings on Naive Methods**

MLP: 2.073

SVM: 1.957

Extra Trees: 1.927

Random Forest: 1.874

Gradient Tree Boosting: 1.869

Kernel Ridge: 1.839

AdaBoost: 1.820